smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/scheduler/SmallFileScheduler.java

```diff
@@ -0,0 +1,492 @@
+      String smallFiles = actionInfo.getArgs().get(HdfsAction.FILE_PATH);
+      ArrayList<String> smallFileList = new Gson().fromJson(
+          smallFiles, new TypeToken<ArrayList<String>>() {
+          }.getType());
+
+      // Check if container file is locked and retry
+      if (containerFileLock.containsKey(containerFilePath)) {
+        return ScheduleResult.RETRY;
+      } else {
+        // Check if small file path is valid and unlocked
+        for (String smallFile : smallFileList) {
+          if (smallFile == null || smallFile.isEmpty()) {
+            LOG.error("Illegal small file path: " + smallFile);
+            actionInfo.setResult("Illegal small file path: " + smallFile);
+            return ScheduleResult.FAIL;
+          } else if (containerFileCache.contains(smallFile)) {
+            LOG.error("This file is not small file: " + smallFile);
+            actionInfo.setResult("This file is not small file: " + smallFile);
+            return ScheduleResult.FAIL;
+          }
+        }
+
+        // Get file state of small files from meta store.
+        Map<String, FileState> fileStateMap;
+        try {
+          fileStateMap = metaStore.getFileStates(smallFileList);
+        } catch (MetaStoreException e) {
+          LOG.error("Failed to get file states of small files. " + e.toString());
+          actionInfo.setResult(
+              "Failed to get file states of small files. " + e.toString());
+          return ScheduleResult.FAIL;
+        }
+
+        // Check if small file is NORMAL
+        if (fileStateMap.size() != 0) {
+          for (String smallFile : smallFileList) {
+            if (fileStateMap.containsKey(smallFile)) {
+              FileState.FileType smallFileType = fileStateMap.get(
+                  smallFile).getFileType();
+              if (smallFileType != FileState.FileType.NORMAL) {
+                String errMsg = String.format(
+                    "%s has invalid file state %s for small file compact.",
+                    smallFile, smallFileType.toString());
+                LOG.error(errMsg);
+                actionInfo.setResult(errMsg);
+                return ScheduleResult.FAIL;
+              }
+            }
+          }
+        }
+
+        try {
+          // Lock container file and set whether exists already
+          // Lock small files
+          boolean exist = dfsClient.exists(containerFilePath);
+          containerFileLock.put(containerFilePath, exist);
+          compactSmallFileLock.addAll(smallFileList);
+        } catch (IOException e) {
+          String errMsg = String.format(
+              "Failed to check if the container file is exists: %s for %s.",
+              containerFilePath, e.toString());
+          LOG.error(errMsg);
+          actionInfo.setResult(errMsg);
+          return ScheduleResult.FAIL;
+        }
+      }
+
+      return ScheduleResult.SUCCESS;
+    } else if (UNCOMPACT_ACTION_NAME.equals(actionInfo.getActionName())) {
```

Comment:
Better to split this function into two: one for `compact` and one for `uncompact`.